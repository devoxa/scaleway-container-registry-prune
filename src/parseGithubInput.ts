const UUID_REGEX = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-5][0-9a-f]{3}-[089ab][0-9a-f]{3}-[0-9a-f]{12}$/i
const IMAGE_REGEX = /^rg\.(.*?)\.scw.cloud\/[^/]+\/([^/]+)$/i

export interface GithubInput {
  scwSecretToken: string
  image: string
  tagPattern: string
  keepLast: string
}

export interface Options {
  scwSecretToken: string
  region: string
  imageName: string
  tagPattern: RegExp
  keepLast: number
}

export function parseGithubInput(input: GithubInput): Options {
  if (!input.scwSecretToken.match(UUID_REGEX)) {
    throw new Error('\'scw-secret-token\' is not a valid "secret key" part of a Scaleway API key')
  }

  const imageMatch = input.image.match(IMAGE_REGEX)
  if (!imageMatch) {
    throw new Error("'image' is not the name of an image on the Scaleway Container Registry")
  }
  const region = imageMatch[1]
  const imageName = imageMatch[2]

  let tagPattern
  try {
    tagPattern = new RegExp(input.tagPattern)
  } catch (err) {
    if (!(err instanceof Error)) throw err
    throw new Error("'tag-pattern' is not a valid regular expression")
  }

  if (!input.keepLast.match(/^\d+$/)) {
    throw new Error("'keep-last' is not a valid positive integer")
  }

  const keepLast = parseInt(input.keepLast, 10)
  if (keepLast >= 100) {
    throw new Error("'keep-last' must be smaller than 100")
  }

  return {
    scwSecretToken: input.scwSecretToken,
    region,
    imageName,
    tagPattern,
    keepLast,
  }
}
