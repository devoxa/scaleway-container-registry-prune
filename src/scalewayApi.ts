import querystring from 'querystring'

type ScwGetImagesResponse = { images?: Array<ScwImage> }
type ScwImage = { id: string; name: string }

type ScwGetTagsResponse = { tags?: Array<ScwTag> }
type ScwTag = { id: string; name: string; updated_at: string }

type ScwDeleteTagResponse = { id: string; name: string; updated_at: string }

export async function getImage(
  scwSecretToken: string,
  region: string,
  imageName: string
): Promise<ScwImage> {
  const response = await fetch(
    `https://api.scaleway.com/registry/v1/regions/${region}/images?name=${imageName}`,
    { headers: { 'x-auth-token': scwSecretToken } }
  )

  const json = (await response.json()) as ScwGetImagesResponse

  if (!json.images || !json.images[0]) {
    throw new Error('Could not get image from the API')
  }

  return json.images[0]
}

export async function listTags(
  scwSecretToken: string,
  region: string,
  imageId: string
): Promise<Array<ScwTag>> {
  const query = querystring.encode({
    page_size: 100,
  })

  const response = await fetch(
    `https://api.scaleway.com/registry/v1/regions/${region}/images/${imageId}/tags?${query}`,
    { headers: { 'x-auth-token': scwSecretToken } }
  )

  const json = (await response.json()) as ScwGetTagsResponse

  if (!json.tags || json.tags.length === 0) {
    throw new Error('Could not list tags from the API')
  }

  // Sort the tags by the time they were last updated (most recent first)
  const tags = json.tags
  tags.sort((a, b) => new Date(b.updated_at).getTime() - new Date(a.updated_at).getTime())

  return tags
}

export async function deleteTag(
  scwSecretToken: string,
  region: string,
  tagId: string
): Promise<ScwDeleteTagResponse> {
  const query = querystring.encode({
    force: true,
  })

  const response = await fetch(
    `https://api.scaleway.com/registry/v1/regions/${region}/tags/${tagId}?${query}`,
    { method: 'DELETE', headers: { 'x-auth-token': scwSecretToken } }
  )

  const json = (await response.json()) as ScwDeleteTagResponse

  if (json.id !== tagId) {
    throw new Error('Could not delete tag from the API')
  }

  return json
}
