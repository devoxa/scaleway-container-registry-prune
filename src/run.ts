import { GithubInput, parseGithubInput } from './parseGithubInput'
import { deleteTag, getImage, listTags } from './scalewayApi'

export async function run(input: GithubInput): Promise<void> {
  const options = parseGithubInput(input)

  const image = await getImage(options.scwSecretToken, options.region, options.imageName)
  const tags = await listTags(options.scwSecretToken, options.region, image.id)

  let count = 0
  for (const tag of tags) {
    if (!tag.name.match(options.tagPattern)) {
      console.log(`Skipping tag: ${tag.name} (${tag.updated_at})`)
      continue
    }

    count++

    if (count <= options.keepLast) {
      console.log(`Keeping tag: ${tag.name} (${tag.updated_at})`)
      continue
    }

    console.log(`Pruning tag: ${tag.name} (${tag.updated_at})`)
    await deleteTag(options.scwSecretToken, options.region, tag.id)
  }
}
