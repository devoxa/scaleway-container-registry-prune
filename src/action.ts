import { getInput } from '@actions/core'
import { run } from './run'

run({
  scwSecretToken: getInput('scw-secret-token'),
  image: getInput('image'),
  tagPattern: getInput('tag-pattern'),
  keepLast: getInput('keep-last'),
}).catch((err) => {
  console.log('Error: ' + err.message)
  process.exit(1)
})
