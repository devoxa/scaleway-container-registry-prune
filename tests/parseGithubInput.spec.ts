import { parseGithubInput } from '../src/parseGithubInput'

const VALID_INPUT = {
  scwSecretToken: '03a23c20-d00c-4e80-9166-bf9318c949de',
  image: 'rg.nl-ams.scw.cloud/devoxa/hello-world',
  tagPattern: '^(?!pr-).+$',
  keepLast: '10',
}

describe('parseGithubInput', () => {
  it('can parse valid input', () => {
    const options = parseGithubInput(VALID_INPUT)
    expect(options).toMatchSnapshot()
  })

  it('errors for invalid scwSecretToken', () => {
    expect(() =>
      parseGithubInput({ ...VALID_INPUT, scwSecretToken: 'foobar' })
    ).toThrowErrorMatchingSnapshot()
  })

  it('errors for invalid image', () => {
    expect(() =>
      parseGithubInput({ ...VALID_INPUT, image: 'scw.cloud/devoxa/hello-world' })
    ).toThrowErrorMatchingSnapshot()
  })

  it('errors for invalid tagPattern', () => {
    expect(() =>
      parseGithubInput({ ...VALID_INPUT, tagPattern: '[' })
    ).toThrowErrorMatchingSnapshot()
  })

  it('errors for invalid keepLast', () => {
    expect(() =>
      parseGithubInput({ ...VALID_INPUT, keepLast: 'TEN' })
    ).toThrowErrorMatchingSnapshot()

    expect(() =>
      parseGithubInput({ ...VALID_INPUT, keepLast: '1000' })
    ).toThrowErrorMatchingSnapshot()
  })
})
