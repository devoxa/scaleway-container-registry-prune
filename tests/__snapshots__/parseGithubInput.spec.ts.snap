// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parseGithubInput can parse valid input 1`] = `
{
  "imageName": "hello-world",
  "keepLast": 10,
  "region": "nl-ams",
  "scwSecretToken": "03a23c20-d00c-4e80-9166-bf9318c949de",
  "tagPattern": /\\^\\(\\?!pr-\\)\\.\\+\\$/,
}
`;

exports[`parseGithubInput errors for invalid image 1`] = `"'image' is not the name of an image on the Scaleway Container Registry"`;

exports[`parseGithubInput errors for invalid keepLast 1`] = `"'keep-last' is not a valid positive integer"`;

exports[`parseGithubInput errors for invalid keepLast 2`] = `"'keep-last' must be smaller than 100"`;

exports[`parseGithubInput errors for invalid scwSecretToken 1`] = `"'scw-secret-token' is not a valid "secret key" part of a Scaleway API key"`;

exports[`parseGithubInput errors for invalid tagPattern 1`] = `"'tag-pattern' is not a valid regular expression"`;
